### PR TITLE
feat: add deployment URL

### PR DESCRIPTION
**Pull request by denysvitali**
### Quick description

This PR allows the plug-ins to add a Deployment URL without the URL Service (Horizon). This is useful when a platform already exposes an URL by default (e.g: Cloud Foundry).

In the case of other plug-ins (such as Kubernetes), the Deployment URL can point to the pod's IP address for example.

---

### In-depth description

This PR adds the URL() string method to the Deployment interface. Unfortunately it seems that if you don't have Horizon (the URL service), you can't get an URL for deployments. Depending on the platform, this can either be intended or not (e.g: Kubernetes can't have public deployment URL without creating an Ingress / Service, but could eventually support a .cluster.local URL).

Some platforms do support a custom URL even for non-productive instances (e.g: Cloud Foundry): a deployment can therefore be accessed without an URL service like Horizon.
With this PR, we want to give the possibility to every plugin to provide (or not) a Deployment URL, so that it will show up in the UI if available.

This change requires some changes in all Waypoint plugins since we now require this new URL() method as part of our Deployment interface. Some changes to handle this new option can be added to Waypoint as well, so that it can display the deployment URL in the consoe and UI, if available.

https://github.com/hashicorp/waypoint/blob/main/internal/cli/deployment_create.go#L54

### CODE REVIEW

File: component/component.go [hunk 1]
```diff
@@ -219,7 +219,9 @@ type Artifact interface {
 	Labels() map[string]string
 }
 
-type Deployment interface{}
+type Deployment interface{
+	URL() string
```
Comment (mitchellh) [inline component/component.go:223]: We should make this optional to: (1) prevent breaking existing plugins and (2) prevent an empty string return for most plugins since this is not the most common.

File: proto/plugin.proto [hunk 2]
```diff
@@ -2,7 +2,7 @@ syntax = "proto3";
 
 package hashicorp.waypoint.sdk;
 
-option go_package = ".;proto";
+option go_package = "github.com/hashicorp/waypoint-plugin-sdk;proto";
```
Comment (mitchellh) [inline proto/plugin.proto:5]: No need to change this in this PR. If this is causing real issues, let's open another issue and talk about it. This feels unrelated to the proposed changes here.

File: proto/plugin.proto [hunk 3]
```diff
@@ -578,10 +578,13 @@ message DefaultReleaser {
 }
 
 message Deploy {
+  string url = 1;
```
Comment (mitchellh) [inline proto/plugin.proto:581]: Comment

File: proto/plugin.proto [hunk 4]
```diff
@@ -578,10 +578,13 @@ message DefaultReleaser {
...
   message Resp {
     // result is the resulting opaque data type
     google.protobuf.Any result = 1;
 
+    Deploy deployment = 3;
```
Comment (mitchellh) [inline proto/plugin.proto:586]: Comment

File: proto/unused.proto [hunk 5]
```diff
@@ -2,4 +2,4 @@ syntax = "proto3";
 
 package hashicorp.waypoint.sdk;
 
-option go_package = ".;proto";
+option go_package = "github.com/hashicorp/waypoint-plugin-sdk;proto";
```
Comment (mitchellh) [inline proto/unused.proto:5]: Same as prior feedback:

> No need to change this in this PR. If this is causing real issues, let's open another issue and talk about it. This feels unrelated to the proposed changes here.

**Review by mitchellh — CHANGES_REQUESTED** [review]

File: component/component.go (hunk 1 above)
Comment (denysvitali) [reply component/component.go]: Solved, I'm now using a `DeploymentWithUrl` as you suggested in the Waypoint PR

File: proto/plugin.proto (hunk 2 above)
Comment (denysvitali) [reply proto/plugin.proto]: Fixed

File: proto/plugin.proto (hunk 3 above)
Comment (denysvitali) [reply proto/plugin.proto:586]: Done

File: proto/plugin.proto (hunk 4 above)
Comment (denysvitali) [reply proto/plugin.proto:592]: Done

File: proto/unused.proto (hunk 5 above)
Comment (denysvitali) [reply proto/unused.proto]: You're right, fixed ;)

File: component/component.go [hunk 6]
```diff
@@ -220,6 +220,9 @@ type Artifact interface {
 }
 
 type Deployment interface{}
+type DeploymentWithUrl interface {
```
Comment (evanphx) [inline component/component.go:223]: Can you add a comment before that documenting what it's used for?

**Review by evanphx — CHANGES_REQUESTED** [review]
One minor thing to help with folks using this feature in the future.

**Review by mitchellh — COMMENTED** [review]
I second the comments from Evan. The rest is looking good.

**Comment by denysvitali** [comment]
Friendly ping

File: go.mod [hunk 7]
```diff
@@ -10,7 +10,7 @@ require (
 	github.com/davecgh/go-spew v1.1.1
 	github.com/fatih/color v1.9.0
 	github.com/golang/protobuf v1.4.2
-	github.com/hashicorp/go-argmapper v0.2.0
+	github.com/hashicorp/go-argmapper v0.1.1
```
Comment (evanphx) [inline go.mod:13]: You need to rebase to main again, sorry, so that we don't downgrade argmapper.

File: internal/plugin/platform.go [hunk 8]
```diff
@@ -436,7 +437,13 @@ func (s *platformServer) Deploy(
 		return nil, err
 	}
 
-	result := &proto.Deploy_Resp{Result: encoded}
+	result := &proto.Deploy_Resp{Result: encoded, Deployment: &proto.Deploy{}}
+
+	switch raw.(type) {
+	case component.DeploymentWithUrl:
+		result.Deployment.Url = raw.(component.DeploymentWithUrl).URL()
+	}
```
Comment (evanphx) [inline internal/plugin/platform.go:445]: One other quick thing, could you either use assignment in the switch cond `r := raw.(type)`, use the condition form of the type check, so that we don't have to do the panic-possible check again on 444.

**Review by evanphx — CHANGES_REQUESTED** [review]
Hiya! A couple more quick changes and I'll be sure it goes in the next release.

**Comment by denysvitali** [comment]
It should be okay now :)

File: component/component.go [hunk 9]
```diff
@@ -221,6 +221,20 @@ type Artifact interface {
...
+// with an URL without using any external URL service (like
+// Hashicorp Horizon). This means that the platform that
+// hosts the workload can provide automatically an URL
+// that is publicly (or company-internally)
+// reachable.
+//
+// Example: automatic routing of Kubernetes pods with
+// an operator, Cloud Foundry's URLs, ...
+type DeploymentWithUrl interface {
+	 URL() string
```
Comment (mitchellh) [inline component/component.go:235]: I would embed the `Deployment` interface here (even though I think its empty right now) just to show that we expect this to be additive to a Deployment. Not a merge blocker though...

**Review by mitchellh — APPROVED** [review]

**Review by evanphx — APPROVED** [review]

**Comment by denysvitali** [comment]
Thanks 🎉
